### ISSUE TITLE
Questions during inference for Wikipedia

### ISSUE THREAD

**Issue by AiliAili** [jmhessel/multi-retrieval#7]
Hi there,
for your wiki dataset, I've seen that there are no gold labels for each document, such as (I've shortened this document)

```
[[["images/1/8094.png", null], ["images/12/1
13835.png", null], ["images/10/95720.png", null]], [["Ubuntu (operating system)", null], ["If you update this article please remember to update other Ubuntu derivatives which include (but are not limited to): Ub
untu, Kubuntu, Edubuntu, Xubuntu, Gobuntu, Ubuntu Studio, and Mythbuntu Ubuntu ( ), is a computer operating system based on the Debian Linux distribution.", null],  ["Canonical endorses and provides support for three additional Ubuntu-derived operating systems: Kubuntu, Edubuntu and Ubuntu Server Edition.", null]],  "text/en/3/327504"]
```
During inference, how do you interpret predictions?
From my understanding, during inference, we can get the textspread score and imagespread score. How do we determine which sentence is more likely to match which image?

Thank you very much.

**Comment by jmhessel** [comment]
Hey again @AiliAili ! :)

As described in Sec 6 of the paper, because there are no labels for wikipedia, we are limited to qualitative evaluation (instead of computing, e.g., precision or AUC, as with the other 6 corpora). That being said, one could hand label a small set of them to estimate AUC. Or --- one could find metadata about the images from wikipedia and see how much they overlap with the predicted links.

**Comment by AiliAili** [comment]
Thank you very much for your timely reply.

Yeah, I've read that part multiple times, but I still don't have a clear idea of what you do to select the most probable images for sentences.
I am not aiming to measure the predictions based on precision or recall. Instead, I am trying to match the most probable image for each sentence. (like the one you show in Fig. 5).

What should I do?  What textspread and imagespread score are used for thresholding?
Thank you very much.

**Comment by jmhessel** [comment]
Oh! Gotcha. We use the estimated M hat matrices (which are produced without labels during the training process). In particular, the M hat matrix estimates similarity between all pairs of images/sentences. Then, much like for the AP algorithm, we solve for the optimal assignment of images to sentences, and use that as our predictions.

**Comment by AiliAili** [comment]
Thank you very much for your explanation.

Given what you said, I should use the data as the training set if I want to predict the correspondence of sentences and images in the dataset?

For AP algorithms, after we get the M hat matrics, I think we should use this one (https://github.com/jmhessel/multi-retrieval/blob/master/bipartite_utils.py ). Could you please provide a sample command line to get the optimal assignment of images to sentences, given the M hat matrices.

Sorry, I have so many questions.
Thank you again for your time. : )

**Comment by jmhessel** [comment]
Hey @AiliAili ---

I am swamped at the moment, so I don't have time to make a command line tool.

However, the function in this file

https://github.com/jmhessel/multi-retrieval/blob/master/bipartite_utils.py

should be okay. These lines are the ones that do it:

https://github.com/jmhessel/multi-retrieval/blob/master/train_doc.py#L462-L466

so, you generate the function, then you can call it with a numpy array (at least in the latest version of tf)

It wouldn't be too hard to make a command line tool for this, but I am swamped right now. Would you potentially be interested in adding a `main` function to `bipartite_utils.py` that takes in a numpy array (e.g., cached as a npy file) and outputs the solution?

Jack
